### PR TITLE
StringConcatFactory taint tracking

### PR DESCRIPTION
**Pull request by manuel-alvarez-alvarez**
# What Does This Do
Add taint tracking support for `+` concatenation in JDK >=9 through `StringConcatFactory`

# Motivation
JDKs with support for invoke dynamic use `StringConcatFactory` in order to do String concatenation through the `+` operator, we need to support it in order to do traint tracking.

### CODE REVIEW

**Review by jandro996 — APPROVED** [review]

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/IastModuleImpl.java [hunk 1]
```diff
@@ -210,6 +212,54 @@ public void onStringBuilderToString(
...
+    }
+    if (rangeCount == 0) {
+      return;
+    }
+
+    final Range[] targetRanges = new Range[rangeCount];
+    int offset = 0, rangeIndex = 0;
+    for (int item : recipeOffsets) {
+      if (item < 0) {
+        offset += (-item);
```
Comment (smola) [inline dd-java-agent/agent-iast/src/main/java/com/datadog/iast/IastModuleImpl.java:249]: Interesting way to write `offset -= item`?
Or is it to make it clear that we're adding a positive number (after flipping the sign)?

File: dd-java-agent/instrumentation/java-lang/java-lang-9/src/main/java/datadog/trace/instrumentation/java/lang/invoke/StringConcatFactoryCallSite.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+  /**
+   * Preprocess the recipe and create an array of offsets where for each offset:
+   *
+   * <ul>
+   *   <li><code>offset < 0</code> length of the current recipe chunk with sign changed
+   *   <li><code>offset >= 0</code> index of the argument
+   * </ul>
+   */
+  private static int[] preprocessRecipe(final String recipe, final Object[] constants) {
```
Comment (smola) [inline dd-java-agent/instrumentation/java-lang/java-lang-9/src/main/java/datadog/trace/instrumentation/java/lang/invoke/StringConcatFactoryCallSite.java:78]: I think we considered before moving this to build time? If it is possible, we might want to optimize that in the future (not in this PR).

**Review by smola — APPROVED** [review]

File: dd-java-agent/instrumentation/java-lang/java-lang-9/src/main/java/datadog/trace/instrumentation/java/lang/invoke/StringConcatFactoryCallSite.java (hunk 2 above)
Comment (manuel-alvarez-alvarez) [reply dd-java-agent/instrumentation/java-lang/java-lang-9/src/main/java/datadog/trace/instrumentation/java/lang/invoke/StringConcatFactoryCallSite.java:78]: What do you mean by build time?, this processing happens when an invoke dynamic is found (instrumentation time), and not when doing the invoke dynamic as before (run time). Is that what you meant?

Comment (smola) [reply dd-java-agent/instrumentation/java-lang/java-lang-9/src/main/java/datadog/trace/instrumentation/java/lang/invoke/StringConcatFactoryCallSite.java:78]: Discussed offline. I meant actual build time, but it's not really possible.

**Pull request merged**